sentry/src/main/java/io/sentry/SentryTracer.java

```diff
@@ -216,7 +220,35 @@ public void finish(@Nullable SpanStatus status) {
       if (transactionFinishedCallback != null) {
         transactionFinishedCallback.execute(this);
       }
-      hub.captureTransaction(transaction);
+      hub.captureTransaction(transaction, this.traceState());
+    }
+  }
+
+  @Override
+  public @Nullable TraceState traceState() {
+    if (hub.getOptions().isTraceSampling()) {
+      if (traceState == null) {
+        final AtomicReference<User> userAtomicReference = new AtomicReference<>();
+        hub.configureScope(
+            scope -> {
+              userAtomicReference.set(scope.getUser());
+            });
+        this.traceState = new TraceState(this, userAtomicReference.get(), hub.getOptions());
+      }
+      return this.traceState;
+    } else {
+      return null;
+    }
+  }
+
+  @Override
+  public @Nullable TraceStateHeader toTraceStateHeader() {
```

Comment:
If `TraceState` has a reference to `options`, could we have `toTraceStateHeader` be part of `TraceState` instead and avoid this method?

Reply:
`TraceState` does not have a reference to `options`. We can put it to a transient field, just somehow does not seem right to me as `TraceState` was meant to be just a data container.

Reply:
That's a fair point. I'm just trying to find any way not to add these two public methods here but I'm about to lose hope (the PR is approved anyway so)

Reply:
`SentryTracer` is `@ApiStatus.Internal` already, so it'd be fine to be there.